tpm/signer.go

```diff
@@ -0,0 +1,77 @@
+package tpm
+
+import (
+	"context"
+	"crypto"
+	"errors"
+	"fmt"
+	"io"
+
+	"go.step.sm/crypto/tpm/storage"
+)
+
+// signer implements crypto.Signer backed by a TPM key
+type signer struct {
+	tpm    *TPM
+	key    Key
+	public crypto.PublicKey
+}
+
+func (s *signer) Public() crypto.PublicKey {
+	return s.public
+}
+
+func (s *signer) Sign(rand io.Reader, digest []byte, opts crypto.SignerOpts) (signature []byte, err error) {
+	ctx := context.Background()
+	if err := s.tpm.Open(ctx); err != nil {
+		return nil, fmt.Errorf("failed opening TPM: %w", err)
+	}
+	defer s.tpm.Close(ctx)
+
+	loadedKey, err := s.tpm.attestTPM.LoadKey(s.key.data)
+	if err != nil {
+		return nil, err
+	}
+	defer loadedKey.Close()
+
+	priv, err := loadedKey.Private(s.public)
+	if err != nil {
+		return nil, err
+	}
+
+	var signer crypto.Signer
+	var ok bool
+	if signer, ok = priv.(crypto.Signer); !ok {
+		return nil, fmt.Errorf("failed getting TPM private key %q as crypto.Signer", s.key.name)
+	}
+
+	return signer.Sign(rand, digest, opts)
+}
+
+// GetSigner returns a crypto.Signer for a TPM key identified by name.
+func (t *TPM) GetSigner(ctx context.Context, name string) (crypto.Signer, error) {
+	if err := t.Open(ctx); err != nil {
+		return nil, fmt.Errorf("failed opening TPM: %w", err)
+	}
+	defer t.Close(ctx)
+
+	key, err := t.store.GetKey(name)
+	if err != nil {
+		if errors.Is(err, storage.ErrNotFound) {
+			return nil, fmt.Errorf("failed getting signer for key %q: %w", name, ErrNotFound)
+		}
+		return nil, err
+	}
```

Comment:
Added the check here too. Due to the nature of the package, in `Sign` the check is also still performed, because the key is unloaded from the TPM in the `defer`. So currently every call to `Sign` reloads the key. We could add a method that loads the key once, then can be used for multiple sign operations, with the caller of the method being responsible for unloading the key.